cmd/download-plugins/main.go

```diff
@@ -0,0 +1,164 @@
+package main
+
+import (
+	"context"
+	"crypto/sha256"
+	_ "embed"
+	"encoding/hex"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"aead.dev/minisign"
+	"github.com/google/go-github/v48/github"
+
+	"github.com/bufbuild/plugins/internal/release"
+)
+
+func main() {
+	if err := run(); err != nil {
+		log.Fatalf("failed to download release: %v", err)
+	}
+}
+
+func run() error {
+	var (
+		minisignPublicKey string
+		releaseTag        string
+	)
+	flag.StringVar(&minisignPublicKey, "minisign-public-key", "", "path to minisign public key file")
+	flag.StringVar(&releaseTag, "release-tag", "", "release to download (default: latest release)")
+	flag.Parse()
+
+	if len(flag.Args()) != 1 {
+		_, _ = os.Stderr.WriteString("usage: download-plugins <dir>\n")
+		flag.PrintDefaults()
+		os.Exit(2)
+	}
+	downloadDir := flag.Arg(0)
+
+	ctx := context.Background()
+	client := release.NewClient(ctx)
+
+	publicKey, err := loadPublicKey(minisignPublicKey)
+	if err != nil {
+		return err
+	}
+
+	var githubRelease *github.RepositoryRelease
+	if releaseTag == "" {
+		githubRelease, err = client.GetLatestRelease(ctx, release.GithubOwnerBufbuild, release.GithubRepoPlugins)
+		if err != nil {
+			return err
+		}
+	} else {
+		githubRelease, err = client.GetReleaseByTag(ctx, release.GithubOwnerBufbuild, release.GithubRepoPlugins, releaseTag)
+		if err != nil {
+			return err
+		}
+	}
+
+	pluginReleases, err := client.LoadPluginReleases(ctx, githubRelease, publicKey)
+	if err != nil {
+		return err
+	}
+	for _, pluginRelease := range pluginReleases.Releases {
+		exists, err := pluginExistsMatchingDigest(pluginRelease, downloadDir)
+		if err != nil {
+			return err
+		}
+		if exists {
+			log.Printf("already downloaded: %s", filepath.Join(downloadDir, filepath.Base(pluginRelease.URL)))
+		} else if err := downloadReleaseToDir(ctx, client.GitHub.Client(), pluginRelease, downloadDir); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func pluginExistsMatchingDigest(plugin release.PluginRelease, downloadDir string) (bool, error) {
+	digest, err := release.CalculateDigest(filepath.Join(downloadDir, filepath.Base(plugin.URL)))
+	if err != nil {
+		return false, err
+	}
+	return digest == plugin.PluginZipDigest, nil
+}
+
+func downloadReleaseToDir(ctx context.Context, client *http.Client, plugin release.PluginRelease, downloadDir string) error {
+	expectedDigest, err := parseDigest(plugin.PluginZipDigest)
+	if err != nil {
+		return err
+	}
+	f, err := os.CreateTemp(downloadDir, "."+strings.ReplaceAll(plugin.PluginName, "/", "-"))
```

Comment:
Downloads the zip file to a temporary file and then renames it when complete (if digest matches).